components/gitpod-db/src/typeorm/workspace-db-impl.ts

```diff
@@ -760,6 +760,30 @@ export abstract class AbstractTypeORMWorkspaceDBImpl implements WorkspaceDB {
         await volumeSnapshots.update(volumeSnapshot.id, volumeSnapshot);
     }
 
+    // finds all workspaces which have more then one volume snapshots
+    // as we only want to keep one latest volume snapshot per workspace
+    public async findVolumeSnapshotWorkspacesForGC(): Promise<string[]> {
+        const volumeSnapshotRepo = await this.getVolumeSnapshotRepo();
+        const qb = volumeSnapshotRepo
+            .createQueryBuilder("vs")
+            .select("vs.workspaceId", "workspaceId")
+            .groupBy("vs.workspaceId")
+            .having("COUNT(*) > 1");
+        const results = (await qb.getRawMany()) as Pick<DBVolumeSnapshot, "workspaceId">[];
+        return results.map((vs) => vs.workspaceId);
+    }
+
+    // returns the list of all volume snapshots for specific workspace id
+    public async findVolumeSnapshotForGCByWorkspaceId(wsId: string): Promise<VolumeSnapshot[]> {
+        const volumeSnapshotRepo = await this.getVolumeSnapshotRepo();
+        const results = await volumeSnapshotRepo
+            .createQueryBuilder("vs")
+            .where("vs.workspaceId = :wsId", { wsId })
```

Comment:
Checking this query it again it might be necessary to instead of introduce `ind_workspaceId`, make it a combined 2ndary key: `ind_workspaceId_creationTime (workspaceId, creationTime)`. This we guarantee it fits here, but can also be used for workspaceId alone. :+1:

Reply:
Fixed.